### PR TITLE
Nifi integration

### PR DESCRIPTION
**Pull request by bputt-e**
We submitted a draft PR where we have a lot of our code commented out because we wanted to trim the issue down to a class not found error.  We need the community's help in understanding the library loader as the BinFiles class is not found when it should be. We added a comment pointing to the particular file/line and have some instructions below on debugging nifi to reproduce the issue.

Download NiFi `https://www.apache.org/dyn/closer.lua?path=/nifi/1.24.0/nifi-1.24.0-bin.zip`

Modify the `conf/bootstrap.conf` in the nifi folder to add some java args so the agent can be loaded

```
java.arg.54=-Dotel.instrumentation.common.default-enabled=false
java.arg.55=-Dotel.instrumentation.nifi.enabled=true
java.arg.56=-javaagent:$POJECT_PATH/opentelemetry-java-instrumentation/javaagent/build/libs/opentelemetry-javaagent-<WHATEVER_VERSION>-SNAPSHOT.jar
```

Start nifi: `bin/nifi.sh run`
Tail nifi logs: `tail -f log/nifi-bootstrap.log`

You'll notice that the nifi instrumentation is disabled because it gets a class not found error for `org.apache.nifi.processor.util.bin.BinFiles`

```
2024-01-18 17:43:58,498 ERROR
[otel.javaagent 2024-01-18 17:43:58:498 +0000] 
[Timer-Driven Process Thread-1] WARN io.opentelemetry.javaagent.tooling.instrumentation.MuzzleMatcher - 
-- io.opentelemetry.javaagent.shaded.instrumentation.nifi.v1_24_0.model.StandardIdentifier:146 Missing class org.apache.nifi.processor.util.bin.BinFiles
```

### CODE REVIEW

File: instrumentation/apache-nifi/apache-nifi-1.24.0/library/src/main/java/io/opentelemetry/instrumentation/nifi/v1_24_0/model/StandardIdentifier.java [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+           if(object instanceof Processor) {
+
+               Processor processor = (Processor) object;
+               System.out.println(">>>>>> DEBUG IN-BINFILES-METHOD : " + processor.getIdentifier());
+
+               //String x = org.apache.nifi.processor.util.bin.BinFiles.class.getName();
+               
+               // IF UNCOMMENT TWO LINES BELOW, MISSING CLASS OCCURS
+               //String x = org.apache.nifi.processors.standard.LogAttribute.class.getName();
```
Comment (bputt-e) [inline instrumentation/apache-nifi/apache-nifi-1.24.0/library/src/main/java/io/opentelemetry/instrumentation/nifi/v1_24_0/model/StandardIdentifier.java:148]: Need community's help in understanding why uncommenting these lines causes a class not found error when we include it in the build.gradle.kts

File: instrumentation/apache-nifi/apache-nifi-1.24.0/library/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+dependencies {
+
+  library("org.apache.nifi:nifi-bin-manager:1.24.0")
+  library("org.apache.nifi:nifi-framework-core:1.24.0")
+  library("org.apache.nifi:nifi-standard-processors:1.24.0")
+  library("javax.servlet:javax.servlet-api:3.1.0")
+  library("com.squareup.okhttp3:okhttp:4.10.0")
+
+  compileOnly(project(":muzzle"))
+
```
Comment (trask) [inline instrumentation/apache-nifi/apache-nifi-1.24.0/library/build.gradle.kts:14]: don't need muzzle in library instrumentation
```suggestion
  compileOnly(project(":muzzle"))

```

File: instrumentation/apache-nifi/apache-nifi-1.24.0/library/build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  id("otel.library-instrumentation")
+}
+
+dependencies {
+
+  library("org.apache.nifi:nifi-bin-manager:1.24.0")
+  library("org.apache.nifi:nifi-framework-core:1.24.0")
+  library("org.apache.nifi:nifi-standard-processors:1.24.0")
+  library("javax.servlet:javax.servlet-api:3.1.0")
+  library("com.squareup.okhttp3:okhttp:4.10.0")
```
Comment (trask) [inline instrumentation/apache-nifi/apache-nifi-1.24.0/library/build.gradle.kts:11]: I think these should be `compileOnly`

**Comment by laurit** [comment]
Conceptually instrumentation has the same effect as modifying the class instrumentation transforms and copying all the dependencies into the same module where the transformed class is. So if you try to modify `org.apache.nifi.controller.repository.StandardProcessSession` that is in `./work/nar/framework/nifi-framework-nar-1.24.0.nar-unpacked/NAR-INF/bundled-dependencies/nifi-framework-components-1.24.0.jar` and inside that class try to use `org.apache.nifi.processors.standard.LogAttribute` that is in `./work/nar/extensions/nifi-standard-nar-1.24.0.nar-unpacked/NAR-INF/bundled-dependencies/nifi-standard-processors-1.24.0.jar` would that work or would you get a compilation error? My guess would be that it won’t work. If you build your library instrumentation how would you hook that up? If you have that figured out then try to do the same thing with the agent. My guess is that instrumenting the `StandardProcessSession` just isn’t the correct strategy, the class loader that loads that class does not have access to the classes you wish to use in the instrumentation.

**Comment by bputt-e** [comment]
@laurit  @trask as mentioned during Thursday's meeting, if there is no expectation of this actually making it into this baseline as an integration, we need to know sooner than later. Can we get commitment from the java instrumentation decision makers on a yes OR no before we spend too much more time going down this path. We're looking into other ways to achieve tracing in Apache NiFi knowing a little more about the class loader issue, but don't want to invest much more time into an integration IF there's no chance it'd actually get pulled into the baseline

**Comment by trask** [comment]
hi @bputt-e, even if it doesn't go into the core Java agent, it can always be published as a OpenTelemetry Java agent extension via https://github.com/open-telemetry/opentelemetry-java-contrib (where your team can be the component owners and have more freedom)

**Comment by bputt-e** [comment]
nifi-javaagent-dependencies

Run the htmlDependencyReport against the the apache-nifi project to verify the existence of the dependencies classpath at runtime. The screenshot above shows the packages that we use but the muzzler says they dont exist

**Pull request closed without merging**